### PR TITLE
Add Pi0.5 model 

### PR DESCRIPTION
**Pull request by changh95**
### Summary

- Adds Physical Intelligence's Pi0.5 VLA model to our list, due to customer request.
https://github.com/tenstorrent/tt-metal/issues/41430

- Not a functional model. Went through quite some optimization.

- PCC is over 99%, and verified e2e model accuracy and performance via Mujoco robotics simulator + LIBERO dataset

- Currently targetd for Blackhole p150 chip.

Refer to 'https://github.com/tenstorrent/tt-metal/tree/changh95/pi05/models/experimental/pi0_5' for proper summary

### Results

Verified on Mujoco + LIBERO

image

### TTNN (Blackhole) vs PyTorch (CPU)

| Metric | TTNN (Blackhole) | PyTorch (CPU) | Speedup |
|---|---|---|---|
| Inference latency (p50) | **163.8 ms** | 5,357 ms | **33×** |
| Inference FPS | **5.78 replans/s** | 0.19 replans/s | **30×** |
| Overall FPS (wall clock) | **31.1 env-steps/s** | 1.7 env-steps/s | **18×** |
| Task success rate | **4/5 (80%)** | 4/5 (80%) | matched |

| PCC    | Latency  | Throughput | Notes |
|--------|----------|------------|-------|
| 0.9977 | 183.4 ms | 272.7 a/s  | baseline, TTNN 0.67.4 wheel |
| 0.9967 | 169.2 ms | 295.5 a/s  | bfloat8_b SigLIP + pre-baked adaRMS |
| 0.9933 | 166.3 ms | 300.6 a/s  | source-built tt-metal v0.65.1rc17 |
| 0.9933 | 151.4 ms | 330.3 a/s  | pre-allocated KV cache (no first-call concat) |
| 0.9921 | 145.6 ms | 343.4 a/s  | KV cache in L1 (SDPA 75µs → 53µs) |
| 0.9921 | 144.6 ms | 345.5 a/s  | fused `rotary_embedding_to_cache` op |
| 0.9921 | 144.1 ms | 347.0 a/s  | precomputed per-step adarms_cond + cached suffix mask |
| **0.9921** | **132.7 ms** | **376.7 a/s** | precomputed per-(step,layer) adaRMS modulations in DRAM |

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:changh95/pi05)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:changh95/pi05)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:changh95/pi05)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:changh95/pi05)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:changh95/pi05)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:changh95/pi05)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:changh95/pi05)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:changh95/pi05)

### CODE REVIEW

**Review by ayerofieiev-tt — APPROVED** [review]

File: models/experimental/pi0_5/common/__init__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
```
Comment (bbradelTT) [inline models/experimental/pi0_5/common/__init__.py]: 2026 USA, Inc. here and elsewhere for new files.

File: models/experimental/pi0_5/tests/demo/run_libero_demo.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    if len(main_images) < 1:
+        print(f"❌ No LIBERO images found in {LIBERO_IMAGES_DIR}")
+        sys.exit(1)
+
+    config = create_config()
+    image_size = config.siglip_config.image_size
+
+    print(f"\n📸 Loading LIBERO benchmark images:")
+    print(f"   Dataset: HuggingFaceVLA/libero")
+    print(f"   Original size: 256x256, Resized to: {image_size}x{image_size}")
```
Comment (bbradelTT) [inline models/experimental/pi0_5/tests/demo/run_libero_demo.py]: Is there a reason to use print instead of logger?

File: ttnn/cpp/ttnn/operations/experimental/transformer/fused_adarms/fused_adarms_nanobind.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            epsilon (float): Small constant for numerical stability. Default: 1e-6.
+
+        Keyword Args:
+            memory_config (ttnn.MemoryConfig, optional): Output memory config.
+
+        Returns:
+            Tuple[ttnn.Tensor, ttnn.Tensor]: (normed, gate)
+    )doc";
+
+    ttnn::bind_function<"fused_adarms", "ttnn.experimental.">(
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/transformer/fused_adarms/fused_adarms_nanobind.cpp]: Please consider renaming to fused_adaptive_rms

**Review by bbradelTT — APPROVED** [review]

**Comment by yieldthought** [comment]
Codex review:

- [models/experimental/pi0_5/tt/ttnn_pi0_model.py:91](https://github.com/tenstorrent/tt-metal/blob/5738454809c8ddd266ee9ef167eb1106c13d9212/models/experimental/pi0_5/tt/ttnn_pi0_model.py#L91) and [models/experimental/pi0_5/tt/ttnn_pi0_model.py:282](https://github.com/tenstorrent/tt-metal/blob/5738454809c8ddd266ee9ef167eb1106c13d9212/models/experimental/pi0_5/tt/ttnn_pi0_model.py#L282): sample_actions never samples per-call initial noise and is fixed to batch 1. The constructor creates self.x_t_ttnn once with shape [1, action_horizon, action_dim], then every sample_actions call starts from that same tensor instead of using the input batch size and fresh noise like the reference DenoisingModule.sample_actions does. Repeated inference calls therefore reuse stale constructor noise, and any batch_size > 1 input cannot produce a matching batch-sized action tensor. Suggestion: allocate/copy a per-call noise tensor sized from state.shape[0], or make the initial noise tensor an explicit API parameter for deterministic evaluation paths.

- [models/experimental/pi0_5/tt/ttnn_gemma.py:78](https://github.com/tenstorrent/tt-metal/blob/5738454809c8ddd266ee9ef167eb1106c13d9212/models/experimental/pi0_5/tt/ttnn_gemma.py#L78): the fused adaRMS fast path bypasses the fallback's 2D cond reshape. The binding documents cond as either [batch, 1, cond_dim] or [batch, cond_dim], but [ttnn/cpp/ttnn/operations/experimental/transformer/fused_adaptive_rms/fused_adaptive_rms.cpp:42](https://github.com/tenstorrent/tt-metal/blob/5738454809c8ddd266ee9ef167eb1106c13d9212/ttnn/cpp/ttnn/operations/experimental/transformer/fused_adaptive_rms/fused_adaptive_rms.cpp#L42) slices the modulation with 3D start/end/step vectors. If adarms_norm_ttnn is called with a 2D conditioning tensor while the fused op is present, it no longer matches the fallback path and will fail or slice the wrong rank. Suggestion: normalize cond to rank 3 before selecting the fused path, or make the C++ op validate/support only rank-3 cond and update the public docs.

- [models/experimental/pi0_5/tt/ttnn_prefix.py:57](https://github.com/tenstorrent/tt-metal/blob/5738454809c8ddd266ee9ef167eb1106c13d9212/models/experimental/pi0_5/tt/ttnn_prefix.py#L57) and [models/experimental/pi0_5/tt/ttnn_prefix.py:187](https://github.com/tenstorrent/tt-metal/blob/5738454809c8ddd266ee9ef167eb1106c13d9212/models/experimental/pi0_5/tt/ttnn_prefix.py#L187): PrefixEmbeddingTTNN always returns the preallocated (1, 544) prefix_att_masks tensor regardless of the actual concatenated prefix length. The reference implementation builds this mask from the emitted image and language token lengths, and this TTNN method already returns dynamic prefix_embs/prefix_pad_masks. Any caller that uses the returned prefix_att mask with a different language length, image count, or batch size gets a mask whose shape no longer matches the prefix tensors. Suggestion: build/cache the zero mask from prefix_embs.shape[:2], or assert the fixed 2-image/32-token/1-batch contract at the boundary.

**Review by jbaumanTT — APPROVED** [review]

**Comment by changh95** [comment]
Thanks @yieldthought — all three addressed in fe7992b:

1. sample_actions per-call noise & batch (ttnn_pi0_model.py). sample_actions now takes an optional noise argument and sizes the initial noise from state.shape[0], reallocating the cached buffer when the batch changes — so batch_size > 1 now produces a correctly-sized action tensor. The default stays a deterministic cached buffer (a reproducible policy is what we want for closed-loop eval/benchmarking), but it's now batch-correct and explicitly overridable, which matches your "explicit API parameter for deterministic evaluation paths" suggestion. The determinism and multi-replan PCC tests now inject noise through this parameter.

2. Fused adaRMS cond rank (ttnn_gemma.py). The fused fast path now reshapes a 2D [batch, cond_dim] cond to rank-3 before calling fused_adaptive_rms, matching the fallback's reshape (the C++ op slices the modulation with rank-3 start/end/step vectors). In this model cond is already rank-3 on the live path, so this closes the latent inconsistency rather than fixing an active failure.

3. Prefix attention mask (ttnn_prefix.py). The zero prefix_att_masks is now built from prefix_embs.shape[:2] (cached by shape) instead of the hardcoded (1, 544), so a different image count / language length / batch gets a correctly-shaped mask instead of a fixed 2-image/32-token/batch-1 tensor.

File: ttnn/cpp/ttnn/operations/experimental/transformer/fused_adaptive_rms/fused_adaptive_rms_nanobind.hpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// SPDX-FileCopyrightText: © 2026 Tenstorrent USA, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <nanobind/nanobind.h>
```
Comment (nsextonTT) [inline ttnn/cpp/ttnn/operations/experimental/transformer/fused_adaptive_rms/fused_adaptive_rms_nanobind.hpp:6]:
```suggestion
#include "ttnn-nanobind/nanobind_fwd.hpp"
```
